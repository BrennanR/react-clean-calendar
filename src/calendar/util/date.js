// @flow

import { arrayRotate } from '../util/array';
import type { Weekday } from '../types';

// https://stackoverflow.com/a/11252167/932896
function treatAsUTC(date: Date): Date {
  const result = new Date(date);
  result.setMinutes(result.getMinutes() - result.getTimezoneOffset());
  return result;
}

/** This function calculates the days between two dates, treating one day as 24 hours and each of the input dates
 * as if they refer to midnight. So startDate: 2018-09-03, endDate: 2018-09-04 would return 1.
 * https://stackoverflow.com/a/11252167/932896
 */
function daysBetween(startDate: Date, endDate: Date): number {
  const millisecondsPerDay = 24 * 60 * 60 * 1000;
  return (treatAsUTC(endDate) - treatAsUTC(startDate)) / millisecondsPerDay;
}

export const dayPerWeekRange = (firstCalendarWeekday: Weekday) =>
  arrayRotate([0, 1, 2, 3, 4, 5, 6], firstCalendarWeekday);

export function daysInFirstCalendarWeek(firstDayOfTheMonth: Date, firstCalendarWeekday: Weekday): number {
  const indexOfFirstWeekdayOfMonth = firstDayOfTheMonth.getDay(); // getDay is indexed Sun = 0, Sat = 6.
  if (firstCalendarWeekday === undefined || indexOfFirstWeekdayOfMonth === undefined) {
    throw new Error('Invalid state. We need both a first day of the month and first calendar weekday index.');
  }
  if (indexOfFirstWeekdayOfMonth >= firstCalendarWeekday) {
    return 7 - indexOfFirstWeekdayOfMonth + firstCalendarWeekday;
  }
  return firstCalendarWeekday - indexOfFirstWeekdayOfMonth;
}

// Sun = 0, Sat = 6.
export function firstWeekdayInMonth(year: number, month: number): Weekday {
  // We know getDay returns 0-6, so force flow to believe we have a Weekday.
  return ((new Date(year, month - 1, 1).getDay(): any): Weekday);
}

export const adjustedDayOffsetBasedOnFirstCalendarWeekday = (dayOffset: number, firstCalendarWeekday: number) => {
  let adjustedDayOffset = dayOffset;
  if (adjustedDayOffset < firstCalendarWeekday) {
    adjustedDayOffset += 7;
  }
  return adjustedDayOffset;
};

export const monthDayOffsetsByWeekForYearMonth = (
  year: number,
  month: number,
  firstCalendarWeekday: Weekday,
): Array<Array<number>> => {
  const weekdayOfTheFirst = firstWeekdayInMonth(year, month);
  const orderedMonthWeekdays = dayPerWeekRange(firstCalendarWeekday);
  const firstDayOfTheMonthDayOffset = adjustedDayOffsetBasedOnFirstCalendarWeekday(
    orderedMonthWeekdays[0],
    firstCalendarWeekday,
  );
  const adjustedFirstDayOfTheMonthOffset =
    firstDayOfTheMonthDayOffset > weekdayOfTheFirst ? firstDayOfTheMonthDayOffset - 7 : firstDayOfTheMonthDayOffset;
  const firstDayOfTheMonthOffsetFromWeekAndDay = adjustedFirstDayOfTheMonthOffset - weekdayOfTheFirst - 1;
  return [...Array(calendarWeeksInMonth(year, month, firstCalendarWeekday)).keys()].map((_, weekIndex) =>
    orderedMonthWeekdays.map((_, dayIndex) => firstDayOfTheMonthOffsetFromWeekAndDay + weekIndex * 7 + dayIndex),
  );
};

export function calendarWeeksInMonth(year: number, month: number, firstCalendarWeekday: Weekday): number {
  // The incoming month is 1 indexed. Convert it to 0 indexed.
  const firstDayOfMonth = new Date(year, month - 1, 1);
  const lastDayOfPreviousMonth = new Date(year, month - 1, 0);
  // 0 for day gives us the last day of the month. Using month directly gives us the next month due to 1-to-0 indexing.
  const lastDayOfMonth = new Date(year, month, 0);
  // This function uses midnight on the input dates to do its calculation. To get the total days in the month, we need
  // the difference between the last day of the previous month, and the last day of the current month.
  const totalDaysInMonth = daysBetween(lastDayOfPreviousMonth, lastDayOfMonth);
  const daysConsumedByFirstWeek = daysInFirstCalendarWeek(firstDayOfMonth, firstCalendarWeekday);
  const daysInMonthExcludingFirstWeek = totalDaysInMonth - daysConsumedByFirstWeek;
  // To get the remaining weeks in the month, take the days that weren't in the first week and divide by 7. Since
  // we need to include an additional calendar week for partial weeks, 'ceil' the division result. Add 1 to account
  // for the first week of the month.
  return Math.ceil(daysInMonthExcludingFirstWeek / 7) + 1;
}

export function nextMonth(year: number, month: number) {
  if (month === 12) {
    return { year: year + 1, month: 1 };
  }
  return { year, month: month + 1 };
}

export function previousMonth(year: number, month: number) {
  if (month === 1) {
    return { year: year - 1, month: 12 };
  }
  return { year, month: month - 1 };
}
