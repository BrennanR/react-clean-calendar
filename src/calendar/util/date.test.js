// @flow

import { calendarWeeksInMonth, monthDayOffsetsByWeekForYearMonth } from './date';

describe('calendarWeeksInMonth', () => {
  it('calculates the number of weeks correctly with Sunday start of week', () => {
    const result = [2015, 2016, 2017, 2018, 2019, 2020].map(year =>
      [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12].map(
        month => `${year}-${month} days: ${calendarWeeksInMonth(year, month, 0)}`,
      ),
    );
    // Validated the snapshot results online against Google Calendar. Weeks/month match.
    expect(result).toMatchSnapshot();
  });

  it('calculates the number of weeks correctly with Monday start of week', () => {
    const result = [2015, 2016, 2017, 2018, 2019, 2020].map(year =>
      [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12].map(
        month => `${year}-${month} days: ${calendarWeeksInMonth(year, month, 1)}`,
      ),
    );
    // Validated the snapshot results online against timeanddate.com. Weeks/month match.
    expect(result).toMatchSnapshot();
  });
});

const SUNDAY = 0;
const MONDAY = 1;
describe('monthDayOffsetsByWeekForYearMonth', () => {
  it('calculates the correct day offsets for Oct 2018 with Sunday start', () => {
    const result = monthDayOffsetsByWeekForYearMonth(2018, 10, SUNDAY);
    expect(result).toMatchSnapshot();
  });
  it('calculates the correct day offsets for Nov 2014 with Sunday start', () => {
    const result = monthDayOffsetsByWeekForYearMonth(2014, 11, SUNDAY);
    expect(result).toMatchSnapshot();
  });

  it('calculates the correct day offsets for June 2018 with Monday start', () => {
    const result = monthDayOffsetsByWeekForYearMonth(2018, 6, MONDAY);
    expect(result).toMatchSnapshot();
  });
  it('calculates the correct day offsets for February 2015 with Monday start', () => {
    const result = monthDayOffsetsByWeekForYearMonth(2015, 2, MONDAY);
    expect(result).toMatchSnapshot();
  });
});
