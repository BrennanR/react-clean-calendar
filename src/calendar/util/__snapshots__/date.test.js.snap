// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendarWeeksInMonth calculates the number of weeks correctly with Monday start of week 1`] = `
Array [
  Array [
    "2015-1 days: 5",
    "2015-2 days: 5",
    "2015-3 days: 6",
    "2015-4 days: 5",
    "2015-5 days: 5",
    "2015-6 days: 5",
    "2015-7 days: 5",
    "2015-8 days: 6",
    "2015-9 days: 5",
    "2015-10 days: 5",
    "2015-11 days: 6",
    "2015-12 days: 5",
  ],
  Array [
    "2016-1 days: 5",
    "2016-2 days: 5",
    "2016-3 days: 5",
    "2016-4 days: 5",
    "2016-5 days: 6",
    "2016-6 days: 5",
    "2016-7 days: 5",
    "2016-8 days: 5",
    "2016-9 days: 5",
    "2016-10 days: 6",
    "2016-11 days: 5",
    "2016-12 days: 5",
  ],
  Array [
    "2017-1 days: 6",
    "2017-2 days: 5",
    "2017-3 days: 5",
    "2017-4 days: 5",
    "2017-5 days: 5",
    "2017-6 days: 5",
    "2017-7 days: 6",
    "2017-8 days: 5",
    "2017-9 days: 5",
    "2017-10 days: 6",
    "2017-11 days: 5",
    "2017-12 days: 5",
  ],
  Array [
    "2018-1 days: 5",
    "2018-2 days: 5",
    "2018-3 days: 5",
    "2018-4 days: 6",
    "2018-5 days: 5",
    "2018-6 days: 5",
    "2018-7 days: 6",
    "2018-8 days: 5",
    "2018-9 days: 5",
    "2018-10 days: 5",
    "2018-11 days: 5",
    "2018-12 days: 6",
  ],
  Array [
    "2019-1 days: 5",
    "2019-2 days: 5",
    "2019-3 days: 5",
    "2019-4 days: 5",
    "2019-5 days: 5",
    "2019-6 days: 5",
    "2019-7 days: 5",
    "2019-8 days: 5",
    "2019-9 days: 6",
    "2019-10 days: 5",
    "2019-11 days: 5",
    "2019-12 days: 6",
  ],
  Array [
    "2020-1 days: 5",
    "2020-2 days: 5",
    "2020-3 days: 6",
    "2020-4 days: 5",
    "2020-5 days: 5",
    "2020-6 days: 5",
    "2020-7 days: 5",
    "2020-8 days: 6",
    "2020-9 days: 5",
    "2020-10 days: 5",
    "2020-11 days: 6",
    "2020-12 days: 5",
  ],
]
`;

exports[`calendarWeeksInMonth calculates the number of weeks correctly with Sunday start of week 1`] = `
Array [
  Array [
    "2015-1 days: 5",
    "2015-2 days: 4",
    "2015-3 days: 5",
    "2015-4 days: 5",
    "2015-5 days: 6",
    "2015-6 days: 5",
    "2015-7 days: 5",
    "2015-8 days: 6",
    "2015-9 days: 5",
    "2015-10 days: 5",
    "2015-11 days: 5",
    "2015-12 days: 5",
  ],
  Array [
    "2016-1 days: 6",
    "2016-2 days: 5",
    "2016-3 days: 5",
    "2016-4 days: 5",
    "2016-5 days: 5",
    "2016-6 days: 5",
    "2016-7 days: 6",
    "2016-8 days: 5",
    "2016-9 days: 5",
    "2016-10 days: 6",
    "2016-11 days: 5",
    "2016-12 days: 5",
  ],
  Array [
    "2017-1 days: 5",
    "2017-2 days: 5",
    "2017-3 days: 5",
    "2017-4 days: 6",
    "2017-5 days: 5",
    "2017-6 days: 5",
    "2017-7 days: 6",
    "2017-8 days: 5",
    "2017-9 days: 5",
    "2017-10 days: 5",
    "2017-11 days: 5",
    "2017-12 days: 6",
  ],
  Array [
    "2018-1 days: 5",
    "2018-2 days: 5",
    "2018-3 days: 5",
    "2018-4 days: 5",
    "2018-5 days: 5",
    "2018-6 days: 5",
    "2018-7 days: 5",
    "2018-8 days: 5",
    "2018-9 days: 6",
    "2018-10 days: 5",
    "2018-11 days: 5",
    "2018-12 days: 6",
  ],
  Array [
    "2019-1 days: 5",
    "2019-2 days: 5",
    "2019-3 days: 6",
    "2019-4 days: 5",
    "2019-5 days: 5",
    "2019-6 days: 6",
    "2019-7 days: 5",
    "2019-8 days: 5",
    "2019-9 days: 5",
    "2019-10 days: 5",
    "2019-11 days: 5",
    "2019-12 days: 5",
  ],
  Array [
    "2020-1 days: 5",
    "2020-2 days: 5",
    "2020-3 days: 5",
    "2020-4 days: 5",
    "2020-5 days: 6",
    "2020-6 days: 5",
    "2020-7 days: 5",
    "2020-8 days: 6",
    "2020-9 days: 5",
    "2020-10 days: 5",
    "2020-11 days: 5",
    "2020-12 days: 5",
  ],
]
`;

exports[`monthDayOffsetsByWeekForYearMonth calculates the correct day offsets for February 2015 with Monday start 1`] = `
Array [
  Array [
    -5,
    -4,
    -3,
    -2,
    -1,
    0,
    1,
  ],
  Array [
    2,
    3,
    4,
    5,
    6,
    7,
    8,
  ],
  Array [
    9,
    10,
    11,
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
    20,
    21,
    22,
  ],
  Array [
    23,
    24,
    25,
    26,
    27,
    28,
    29,
  ],
]
`;

exports[`monthDayOffsetsByWeekForYearMonth calculates the correct day offsets for June 2018 with Monday start 1`] = `
Array [
  Array [
    -3,
    -2,
    -1,
    0,
    1,
    2,
    3,
  ],
  Array [
    4,
    5,
    6,
    7,
    8,
    9,
    10,
  ],
  Array [
    11,
    12,
    13,
    14,
    15,
    16,
    17,
  ],
  Array [
    18,
    19,
    20,
    21,
    22,
    23,
    24,
  ],
  Array [
    25,
    26,
    27,
    28,
    29,
    30,
    31,
  ],
]
`;

exports[`monthDayOffsetsByWeekForYearMonth calculates the correct day offsets for Nov 2014 with Sunday start 1`] = `
Array [
  Array [
    -5,
    -4,
    -3,
    -2,
    -1,
    0,
    1,
  ],
  Array [
    2,
    3,
    4,
    5,
    6,
    7,
    8,
  ],
  Array [
    9,
    10,
    11,
    12,
    13,
    14,
    15,
  ],
  Array [
    16,
    17,
    18,
    19,
    20,
    21,
    22,
  ],
  Array [
    23,
    24,
    25,
    26,
    27,
    28,
    29,
  ],
  Array [
    30,
    31,
    32,
    33,
    34,
    35,
    36,
  ],
]
`;

exports[`monthDayOffsetsByWeekForYearMonth calculates the correct day offsets for Oct 2018 with Sunday start 1`] = `
Array [
  Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
  ],
  Array [
    7,
    8,
    9,
    10,
    11,
    12,
    13,
  ],
  Array [
    14,
    15,
    16,
    17,
    18,
    19,
    20,
  ],
  Array [
    21,
    22,
    23,
    24,
    25,
    26,
    27,
  ],
  Array [
    28,
    29,
    30,
    31,
    32,
    33,
    34,
  ],
]
`;
